src/trio/_core/_exceptions.py

```diff
@@ -67,11 +73,49 @@ class Cancelled(BaseException, metaclass=NoPublicConstructor):
 
     """
 
+    source: Literal["deadline", "nursery", "explicit", "unknown", "KeyboardInterrupt"]
+    # TODO: this should probably be a Task?
+    source_task: str | None = None
+    reason: str | None = None
+
     def __str__(self) -> str:
-        return "Cancelled"
+        def repr_if_not_none(lead: str, s: str | None, trail: str = "") -> str:
+            if s is None:
+                return ""
+            return lead + s + trail
+
+        return (
+            f"cancelled due to {self.source}"
+            + repr_if_not_none(" with reason '", self.reason, "'")
```

Comment:
```suggestion
            + repr_if_not_none(" with reason ", repr(self.reason))
```
(imagine if the reason contains `'`s?)